src/bloqade/squin/analysis/unitary/impls.py

```diff
@@ -0,0 +1,66 @@
+from kirin import interp
+from kirin.analysis import ForwardFrame, const
+from kirin.dialects import scf, func
+
+from ... import op
+from .lattice import Unitary, NotUnitary
+from .analysis import UnitaryAnalysis
+from ..hermitian import Hermitian, NotHermitian
+
+
+@op.dialect.register(key="squin.unitary")
+class UnitaryMethods(interp.MethodTable):
+    @interp.impl(op.stmts.Control)
+    @interp.impl(op.stmts.Adjoint)
+    def simple_container(
+        self,
+        interp: UnitaryAnalysis,
+        frame: ForwardFrame,
+        stmt: op.stmts.Control | op.stmts.Adjoint,
+    ):
+        return (frame.get(stmt.op),)
+
+    @interp.impl(op.stmts.Scale)
+    def scale(self, interp: UnitaryAnalysis, frame: ForwardFrame, stmt: op.stmts.Scale):
+        is_unitary = frame.get(stmt.op)
+
+        # NOTE: need to check if the factor has absolute value squared of 1
+        constant_value = stmt.factor.hints.get("const")
+        if not isinstance(constant_value, const.Value):
+            return (NotUnitary(),)
+
+        num = constant_value.data
+        if not isinstance(num, (float, int, bool, complex)):
```

Comment:
Huh, just learned that `abs` works on bools but now I have to ask, is there actually a case where somebody will pass in a boolean for the `factor` arg considering we limit it to `NumberType`?

Reply:
Right, I admit I'm splitting hairs here I just think the presence of `bool` itself here implies we support it in some kind of way when we really don't. Still a happy camper here though!